### ISSUE TITLE
Https server hang on `close_fn`

### ISSUE THREAD

**Issue by reinismu** [esp-rs/esp-idf-svc#287]
Moving my investigation here https://github.com/espressif/esp-idf/issues/11738

Summary: If there are multiple threads doing job when https server is running there is a chance it will hang. So far last thing that gets called is `close_fn` callback. It is reproduced if I put load on other thread when https call is made

It hangs at
https://github.com/esp-rs/esp-idf-svc/blob/cf0b85c7d92c7382cd87d5abb6b3a1bede70de6f/src/http/server.rs#L447

Note: Not sure why websocket close handler is even called when I open Https endpoint

I think I found where it locks, but not sure why it would lock.

```rust
pub struct Api {
    ...
    ws_listeners: Arc<Mutex<HashMap<i32, EspHttpWsDetachedSender>>>,
}

            ...
            .ws_handler::<_, WsError>("/ws", move |con| {
                log::info!("My ws handler is called");
                if con.is_new() {
                    let session_id = con.session();
                    log::info!("Ws connection open: {session_id}");
                    let detached_sender = con.create_detached_sender().unwrap();
                    ws_listeners
                        .lock()
                        .unwrap()
                        .insert(session_id, detached_sender);

                    return Ok(());
                }

                if con.is_closed() {
                    let session_id = con.session();
                    log::info!("Ws connection closed: {session_id}");
                    ws_listeners.lock().unwrap().remove(&session_id);
                    log::info!("Ws connection closed unlocked");
                    return Ok(());
                }
                Ok(())
            })?;
            
            ...
            
            let mut listeners = self.ws_listeners.lock().unwrap();
            for (_, sender) in listeners.iter_mut() {
                sender
                    .send(embedded_svc::ws::FrameType::Text(false), frame.as_bytes())
                    .map(|_| log::info!("send: {}", frame))
                    .map_err(|e| log::error!("Error sending ws frame: {:?}", e));
            }
```

Stuck at `ws_listeners.lock().unwrap().remove(&session_id);`. I'm not sure how it managed to deadlock tho

**Comment by reinismu** [comment]
Has anyone used websockets where messages are sent out from another thread? Still not sure why it locks, but I wonder how I should change the logic for it to avoid locks. I could use `try_lock`, but then I risk of going into bad state

**Comment by reinismu** [comment]
Thanks for explanation! 

Now I use `try_lock` for `ws_listeners.lock().unwrap().remove(&session_id);` it is mostly called from http call so it doesn't remove anything anyway. Hope this will be fine until a more mature async HTTP server comes along. 

A while back I implemented whole async websocket stack

```
smol = "1.2"
async-tungstenite = { version = "0.19", features = ["async-tls", "async-std-runtime"] }
futures = "0.3"
tungstenite = { version = "0.18", features = ["rustls-tls-native-roots"] }
async-tls = "0.12.0"
rustls = "0.20.8"
```

It was tricky and I even had to make `https://github.com/briansmith/ring` compile for esp32s3. In the end I got async WS working, but rustls was just too resource-heavy for the chip, so I dropped the idea.
